00X-ABI-Stable-Module-API.md

```diff
@@ -0,0 +1,377 @@
+  and inspection.
+* Add in VM agnostic methods for object lifecycle management.
+* Add in VM agnostic exception management
+* Define C/C++ types for the API which are independent from V8
+  (NAN currently uses the V8 types).
+* Use the NAN examples to work through the API and build up the
+  definition of the key methods.
+* Validate on most used modules in the ecosystem
+* Complete performance analysis to confirm accepted level of overhead.
+
+Notes/details on other related work and the current experimental progress
+on implementing this eps can be found in:
+https://docs.google.com/document/d/1bX9SZKufbfhr7ncXL9fFev2LrljymtFZdWpwmv7mB9g/edit#heading=h.ln9aj9qjo969
+
+# API definition
+
+**WORK IN PROGRESS, DRIVEN BY MODULES PORTED SO FAR**
+
+This API will be built up from a minimal set. Based on this set
+we have been able to port the NaN example and level down.  Additions
+will be made as we port additional modules.
+
+### Types
+```C
+typedef struct napi_env__ *napi_env;
+typedef struct napi_value__ *napi_value;
+typedef struct napi_persistent__ *napi_persistent;
+typedef struct napi_handle_scope__ *napi_handle_scope;
+typedef struct napi_escapable_handle_scope__ *napi_escapable_handle_scope;
+typedef struct napi_propertyname__ *napi_propertyname;
+typedef const struct napi_func_cb_info__ *napi_func_cb_info;
+typedef void (*napi_callback)(napi_env, napi_func_cb_info);
+typedef void napi_destruct(void* v);
+
+
+struct napi_method_descriptor {
+  napi_callback callback;
+  char* utf8name;
+};
+
+extern "C" {
+enum napi_valuetype {
+    // ES6 types (corresponds to typeof)
+    napi_undefined,
+    napi_null,
+    napi_boolean,
+    napi_number,
+    napi_string,
+    napi_symbol,
+    napi_object,
+    napi_function,
+};
+```
+### Functions
+
+```C
+// Getters for defined singletons
+NODE_EXTERN napi_value napi_get_undefined(napi_env e);
+NODE_EXTERN napi_value napi_get_null(napi_env e);
+NODE_EXTERN napi_value napi_get_false(napi_env e);
+NODE_EXTERN napi_value napi_get_true(napi_env e);
+NODE_EXTERN napi_value napi_get_global_scope(napi_env e);
+
+
+// Methods to create Primitive types/Objects
+NODE_EXTERN napi_value napi_create_object(napi_env e);
+NODE_EXTERN napi_value napi_create_array(napi_env e);
+NODE_EXTERN napi_value napi_create_array_with_length(napi_env e, int length);
+NODE_EXTERN napi_value napi_create_number(napi_env e, double val);
+NODE_EXTERN napi_value napi_create_string(napi_env e, const char*);
+NODE_EXTERN napi_value napi_create_string_with_length(napi_env e, const char*,
+                                                               size_t length);
+NODE_EXTERN napi_value napi_create_boolean(napi_env e, bool b);
+NODE_EXTERN napi_value napi_create_symbol(napi_env e, const char* s = NULL);
+NODE_EXTERN napi_value napi_create_function(napi_env e, napi_callback cbinfo);
+NODE_EXTERN napi_value napi_create_error(napi_env e, napi_value msg);
+NODE_EXTERN napi_value napi_create_type_error(napi_env e, napi_value msg);
```

Comment:
How will these functions signal error?  E.g. what if `napi_create_object()` can't create an object because the VM is out of memory or in a state where it can't be entered?